File: java/src/main/java/io/fivexer/sdk/WorkerAttachments.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.Attachment;
import io.fivexer.sdk.model.AttachmentDownload;
import io.fivexer.sdk.model.AttachmentEnvelope;
import io.fivexer.sdk.model.AttachmentList;
import io.fivexer.sdk.model.CreatedAttachment;
import io.fivexer.sdk.model.WorkerCreateAttachment;
import java.util.Collections;
import java.util.List;

/**
 * Files on the worker's own tasks — how an unattended agent hands over a deliverable as a file
 * instead of a chunked comment thread.
 *
 * <p>The same storage core as {@link TaskAttachments}: bytes go straight to object storage via a
 * presigned PUT, and {@link #confirm} is what makes them readable. Two differences, both
 * deliberate. The uploader is derived from the session, so {@link WorkerCreateAttachment} has no
 * {@code workerId} to spoof. And there is no {@code remove} — files on a task are an operator's
 * to manage and a worker's only to add and read; a delete here would let a worker erase the
 * evidence of their own work.
 *
 * <p>Deployments without object storage answer 501 {@code storage_unavailable}.
 */
public final class WorkerAttachments {

    private final FivexerWorker client;

    WorkerAttachments(FivexerWorker client) {
        this.client = client;
    }

    /** Reserve an attachment record and get a presigned URL to PUT the bytes to. */
    public CreatedAttachment create(String taskId, WorkerCreateAttachment input) {
        return client.request("POST", "/portal/tasks/" + Json.enc(taskId) + "/attachments",
                input.toJson(), CreatedAttachment.class, false);
    }

    /** Confirm the bytes landed — the server HEADs the object as the authoritative size check. */
    public Attachment confirm(String taskId, String attachmentId) {
        AttachmentEnvelope envelope = client.request("POST",
                "/portal/tasks/" + Json.enc(taskId) + "/attachments/" + Json.enc(attachmentId) + "/confirm",
                null, AttachmentEnvelope.class, false);
        return envelope.getAttachment();
    }

    public List<Attachment> list(String taskId) {
        AttachmentList envelope = client.request("GET",
                "/portal/tasks/" + Json.enc(taskId) + "/attachments", null, AttachmentList.class, false);
        return envelope.getAttachments() == null ? Collections.emptyList() : envelope.getAttachments();
    }

    /** A short-lived presigned download URL for a confirmed attachment. */
    public AttachmentDownload download(String taskId, String attachmentId) {
        return client.request("GET",
                "/portal/tasks/" + Json.enc(taskId) + "/attachments/" + Json.enc(attachmentId) + "/download",
                null, AttachmentDownload.class, false);
    }

    /**
     * Create, PUT the bytes to object storage, then confirm — in one call.
     *
     * <p>The size is derived from {@code payload}, and the presigned headers are sent verbatim
     * because they are part of the signature. A non-2xx from storage throws a
     * {@link FivexerApiException} with code {@code upload_failed}, leaving the record unconfirmed
     * rather than claiming bytes that never landed.
     */
    public Attachment upload(String taskId, byte[] payload, String filename, String contentType) {
        CreatedAttachment created = create(taskId,
                new WorkerCreateAttachment(filename, contentType, (long) payload.length));
        client.putBytes(created.getUpload(), payload);
        return confirm(taskId, created.getAttachment().getId());
    }
}

File: php/src/Resource/WorkerAttachments.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\FivexerWorker;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\Attachment;
use Fivexer\SDK\Model\AttachmentDownload;
use Fivexer\SDK\Model\CreatedAttachment;
use Fivexer\SDK\Model\WorkerCreateAttachment;

/**
 * Files on the worker's own tasks — how an unattended agent hands over a deliverable as a file
 * instead of a chunked comment thread.
 *
 * The same storage core as {@see TaskAttachments}: bytes go straight to object storage via a
 * presigned PUT, and {@see self::confirm()} is what makes them readable. Two differences, both
 * deliberate. The uploader is derived from the session, so {@see WorkerCreateAttachment} has no
 * `workerId` to spoof. And there is no `remove` — files on a task are an operator's to manage
 * and a worker's only to add and read; a delete here would let a worker erase the evidence of
 * their own work.
 *
 * Deployments without object storage answer 501 `storage_unavailable`.
 */
final class WorkerAttachments
{
    public function __construct(private readonly FivexerWorker $client)
    {
    }

    /** Reserve an attachment record and get a presigned URL to PUT the bytes to. */
    public function create(string $taskId, WorkerCreateAttachment $input): CreatedAttachment
    {
        $data = $this->client->portalRequest(
            'POST',
            '/portal/tasks/' . \rawurlencode($taskId) . '/attachments',
            $input->toArray()
        ) ?? [];
        return CreatedAttachment::fromArray($data);
    }

    /** Confirm the bytes landed — the server HEADs the object as the authoritative size check. */
    public function confirm(string $taskId, string $attachmentId): Attachment
    {
        $data = $this->client->portalRequest(
            'POST',
            '/portal/tasks/' . \rawurlencode($taskId)
                . '/attachments/' . \rawurlencode($attachmentId) . '/confirm'
        ) ?? [];
        return Attachment::fromArray($data['attachment'] ?? []);
    }

    /** @return list<Attachment> */
    public function list(string $taskId): array
    {
        $data = $this->client->portalRequest(
            'GET',
            '/portal/tasks/' . \rawurlencode($taskId) . '/attachments'
        ) ?? [];
        /** @var list<Attachment> */
        return Json::parseEach($data, 'attachments', [Attachment::class, 'fromArray']);
    }

    /** A short-lived presigned download URL for a confirmed attachment. */
    public function download(string $taskId, string $attachmentId): AttachmentDownload
    {
        $data = $this->client->portalRequest(
            'GET',
            '/portal/tasks/' . \rawurlencode($taskId)
                . '/attachments/' . \rawurlencode($attachmentId) . '/download'
        ) ?? [];
        return AttachmentDownload::fromArray($data);
    }

    /**
     * Create, PUT the bytes to object storage, then confirm — in one call.
     *
     * The size is derived from $payload, and the presigned headers are sent verbatim because
     * they are part of the signature. A non-2xx from storage throws a FivexerApiException with
     * code `upload_failed`, leaving the record unconfirmed rather than claiming bytes that never
     * landed.
     */
    public function upload(string $taskId, string $payload, string $filename, string $contentType): Attachment
    {
        $created = $this->create(
            $taskId,
            new WorkerCreateAttachment($filename, $contentType, \strlen($payload))
        );
        $this->client->putBytes($created->upload, $payload);
        return $this->confirm($taskId, $created->attachment->id);
    }
}
